packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,157 @@
+import { StaticDataView } from '../../data-view';
+import Preprocessor, { Env } from '../../preprocessor';
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView, env: Env): PreprocessorBucket {
+    const preprocessors: Preprocessor[] = [];
+
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      env,
+      preprocessors,
+    });
+  }
+
+  public readonly conditions: Map<string, Preprocessor>;
+  public readonly ineligible: Set<number>;
+
+  public env: Env;
+
+  constructor({ env, preprocessors = [] }: { env: Env; preprocessors?: Preprocessor[] }) {
+    this.env = env;
+
+    this.ineligible = new Set();
+    this.conditions = new Map();
+
+    this.update({ added: preprocessors });
+  }
+
+  private build() {
+    // Build filters to preprocessor set
+    // We'll use `Set` instead of `Array` to effectively remove duplicates
+    const bindings = new Map<number, Set<Preprocessor>>();
+
+    for (const preprocessor of this.conditions.values()) {
+      // Remove unused preprocessor
+      if (!(preprocessor.positives.size + preprocessor.negatives.size)) {
+        this.conditions.delete(preprocessor.condition);
+        continue;
+      }
+
+      for (const positive of preprocessor.positives) {
+        const binding = bindings.get(positive) ?? new Set();
+
+        binding.add(preprocessor);
+        bindings.set(positive, binding);
+      }
+
+      for (const negative of preprocessor.negatives) {
```

Comment:
not sure why we have to keep the list of positive and negative separately. don't we handle them the same?

Reply:
Each can be renamed into `filters` and `negativeFilters`. If we don't want to use separated list, there are two ways to express negative block.

1) Use `NegatedPreprocessor` class and reference `Preprocessor`.

`NegatedPreprocessor` was used to contain filter ids in `!#else` block. This class doesn't have any independent data structure because it's not efficient to save duplicate conditions in multiple preprocessor objects.

2) Move `positives` into separate data structure and use `negatives` as an negated filter ids.

In this case, we're bringing the data structure from `private update` function of `PreprocessorBucket` to the `PreprocessorBucket`. In this case, we're able to hold a list of preprocessors applied to specific filter id. It means we can save the information related to `positives` in `Map<number, Preprocessor[]>`.

Reply:
Resolved by implementing parenthesis.